server/src/openfga-yaml-schema.ts

```diff
@@ -1,4 +1,348 @@
+  };
+};
+
+
+// Format enforcement
+
+function formatUser(user: string): boolean {
+  if (!user.match(new RegExp(`^${identifier}:(\\*|${identifier}(#${identifier})?)$`)) || user.length > 512) {
+    return false;
+  }
+  return true;
+}
+
+function formatRelation(relation: string): boolean {
+  if (!relation.match(new RegExp(`^${identifier}$`)) || relation.length > 50) {
+    return false;
+  }
+  return true;
+}
+
+function formatObject(object: string): boolean {
+  if (!object.match(new RegExp(`^${identifier}:${identifier}$`)) || object.length > 256) {
+    return false;
+  }
+  return true;
+}
+
+function formatCondition(condition: string): boolean {
+  if (!condition.match(new RegExp(`^${identifier}$`)) || condition.length > 50) {
+    return false;
+  }
+  return true;
+}
+
+// Validation for Types
+
+function validateTypes(tuple: Tuple, types: string[], instancePath: string): boolean {
+  const errors = [];
+
+  const user = tuple.user.split(":")[0];
+
+  // Ensure valid type of user
+  if (!types.includes(user)) {
+    errors.push(invalidType(user, types, instancePath + "/user"));
+  }
+
+  const object = tuple.object.split(":")[0];
+
+  // Ensure valid type of object
+  if (!types.includes(object)) {
+    errors.push(invalidType(object, types, instancePath + "/object"));
+  }
+
+  // Report all errors
+  if (errors.length) {
+    validateTuple.errors?.push(...errors);
+    return false;
+  }
+  return true;
+}
+
+// Validate Relation
+function validateRelation(
+  tuple: Tuple,
+  typeDefs: TypeDefinition[],
+  instancePath: string): boolean {
+  const errors = [];
+
+  // Check if relation exists on given type
+  let doesExisitOnType = false;
+  if (tuple.user.includes("#")) {
+    const user = tuple.user.split(":")[0];
+    const memberRelation = tuple.user.split("#")[1];
+    for (const t of typeDefs) {
+      if (t !== undefined && t.relations && t.type === user && t.relations[memberRelation]) {
+        doesExisitOnType = true;
+        break;
+      }
+    }
+
+    if (!doesExisitOnType) {
+      errors.push(relationMustExistOnType(memberRelation, user, instancePath + "/user"));
+    }
+  }
+
+  // Check if relation exists on given object
+  const objectType = tuple.object.split(":")[0];
+  let doesExisitOnObject = false;
+  for (const t of typeDefs) {
+    if (t !== undefined && t.relations && t.type === objectType && t.relations[tuple.relation]) {
+      doesExisitOnObject = true;
+      break;
+    }
+  }
+  if (!doesExisitOnObject) {
+    errors.push(relationMustExistOnType(tuple.relation, objectType, instancePath + "/relation"));
+  }
+
+  if (errors.length) {
+    validateTuple.errors?.push(...errors);
+    return false;
+  }
+  return true;
+}
+
+function validateTypeRestrictions(
+  tuple: Tuple,
+  typeDefs: TypeDefinition[],
+  conditions: {[key: string]: Condition} | undefined,
+  instancePath: string): boolean {
+
+  validateTuple.errors = validateTuple.errors || [];
+
+  // Wildcard
+  if (tuple.user.includes(":*")) {
```

Comment:
This seems like a lot of repetition.

Can we maybe abstract the bulk of it?

e.g. map tuple.user to:
{
  type: ...
  relation: ...
  wildcard: ...,
  condition: ...
}

and then later when validating, validate accordingly based on what the user is

Reply:
This was a good suggestion, saved a lot of repetition.